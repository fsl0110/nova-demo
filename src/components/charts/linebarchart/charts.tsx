import React, { PureComponent } from "react";
import produce from "immer";
import { Linechart } from "./linechart";
import { Barchart } from "./barchart";

interface Props {
  linechartData: any;
  barchartData: any;
}
interface State {
  linechartData: { year: number; value: number }[] | undefined;
  barchartData: { month: number; value: number }[] | undefined;
}

// Charts Component
export class Charts extends PureComponent<Props, State> {
  state: State = {
    linechartData: [],
    barchartData: []
  };

  componentWillMount() {
    this.setState(
      produce((draft: State) => {
        draft.barchartData = this.props.barchartData;
        draft.linechartData = this.props.linechartData;
      })
    );
  }

  handleChangeYear = (year: string) =>
    this.setState(
      produce((draft: State) => {
        draft.barchartData = this.props.barchartData[year];
      })
    );

  render() {
    const margins = { top: 50, right: 100, bottom: 50, left: 100 },
      svgDimensions = {
        height: window.screen.height / 2,
        width: window.screen.width / 2
      };
    console.log("#############", this.state.barchartData);
    return (
      <div className="chart">
        <div className="lineChart">
          <Linechart
            margins={margins}
            svgDimensions={svgDimensions}
            data={this.state.linechartData}
            onChangeYear={this.handleChangeYear.bind(this)}
          />
        </div>
        <div className="barChart">
          <Barchart
            margins={margins}
            svgDimensions={svgDimensions}
            data={this.state.barchartData}
          />
        </div>
      </div>
    );
  }
}
