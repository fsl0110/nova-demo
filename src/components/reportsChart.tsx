import React, { PureComponent, HTMLAttributes } from "react";
import { connect } from "react-redux";
import produce from "immer";
import Highcharts from "highcharts/highstock";
import HighchartsReact from "highcharts-react-official";
import moment from "moment";
import classNames from "classnames";
import { Search } from "./search";
import { Error } from "./error";
import { fetchReports } from "../actions";
import { AppState } from "../typings";

require("highcharts-no-data-to-display")(Highcharts);

export interface Props extends HTMLAttributes<HTMLDivElement> {
  /** Is loading data */
  isLoading: boolean;
  /** Has an error */
  hasError: boolean;
  /** Fetched and formatted data**/
  reportsData: [];
  /** Action for fetching data **/
  fetchReports: (term?: string) => void;
}

interface State {
  chartOptions: Highcharts.Options;
}

export class ReportsChart extends PureComponent<Props, State> {
  private chartInstance: any;

  constructor(props: Props) {
    super(props);

    this.afterChartCreated = this.afterChartCreated.bind(this);
    this.state = {
      chartOptions: {
        chart: {
          height: 350,
          style: {
            fontFamily:
              "-apple-system, BlinkMacSystemFont, Segoe UI, Roboto, Oxygen"
          }
        },

        credits: {
          enabled: false
        },
        loading: {
          hideDuration: 500,
          showDuration: 300
        },
        yAxis: {
          title: {
            text: null
          }
        },
        xAxis: {
          type: "datetime",
          labels: {
            formatter: function() {
              return moment(this.value).format("MMM YYYY");
            }
          },
          minTickInterval: moment.duration(1, "month").asMilliseconds()
        },
        series: [{ type: "line", data: [] }]
      }
    };
  }

  afterChartCreated(chart: any) {
    this.chartInstance = chart;
  }

  static getDerivedStateFromProps(nextProps: Props, prevState: State) {
    if (prevState.chartOptions.series![0].data !== nextProps.reportsData) {
      return produce(prevState, draft => {
        draft.chartOptions.series![0].data = nextProps.reportsData.map(
          (i: { time: number; count: number }) => [
            moment(i.time, "YYYY-MM-DD").valueOf(),
            i.count
          ]
        );
      });
    }
    return null;
  }

  componentDidMount() {
    this.props.fetchReports("");
  }

  render() {
    const { isLoading, hasError, fetchReports, className } = this.props;

    if (!hasError && this.chartInstance && isLoading) {
      this.chartInstance.showLoading();
    }
    if (!hasError && this.chartInstance && !isLoading) {
      this.chartInstance.hideLoading();
    }
    return (
      <div className={classNames(className)}>
        <Search fetchData={fetchReports} isLoading={isLoading} />
        <div>
          <h3 className="horizontal-center">Enforcement reports over time</h3>
          <div className="chart-wrapper">
            {hasError ? (
              <HighchartsReact
                highcharts={Highcharts}
                options={this.state.chartOptions}
                constructorType={"stockChart"}
                callback={this.afterChartCreated}
              />
            ) : (
              <Error fetchData={fetchReports} isLoading={isLoading} />
            )}
          </div>
        </div>
      </div>
    );
  }
}

export const ReportsChartConnected = connect(
  (state: AppState) => ({
    isLoading: state.reports.isLoading,
    hasError: state.reports.hasError,
    reportsData: state.reports.data
  }),
  {
    fetchReports
  }
)(ReportsChart);

export default ReportsChart;
