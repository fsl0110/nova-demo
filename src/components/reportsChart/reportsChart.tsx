import React, { PureComponent, HTMLAttributes } from "react";
import { connect } from "react-redux";
import Highcharts from "highcharts/highstock";
import HighchartsReact from "highcharts-react-official";
import classNames from "classnames";
import { Error, Search } from "..";
import { fetchReports } from "../../actions";
import { AppState } from "../../typings";

require("highcharts-no-data-to-display")(Highcharts);

export interface Props extends HTMLAttributes<HTMLDivElement>, DispatchProps {
  /** Is loading data */
  isLoading: boolean;
  /** Has an error */
  hasError: boolean;
  /** Options and Data of the chart */
  chartOptions: Highcharts.Options;
}

export interface DispatchProps {
  /** Action for fetching data */
  fetchReports: (term?: string) => void;
}

export class ReportsChart extends PureComponent<Props> {
  private chartInstance: any;

  constructor(props: Props) {
    super(props);
    this.afterChartCreated = this.afterChartCreated.bind(this);
  }

  afterChartCreated(chart: any) {
    this.chartInstance = chart;
  }

  componentDidMount() {
    this.props.fetchReports("");
  }

  render() {
    const {
      isLoading,
      hasError,
      fetchReports,
      chartOptions,
      className
    } = this.props;

    if (!hasError && this.chartInstance && isLoading) {
      this.chartInstance.showLoading();
    }
    if (!hasError && this.chartInstance && !isLoading) {
      this.chartInstance.hideLoading();
    }
    return (
      <div className={classNames(className)}>
        <Search fetchData={fetchReports} isLoading={isLoading} />
        <div>
          <h3 className="horizontal-center">Enforcement reports over time</h3>
          <div className="chart-wrapper">
            {!hasError ? (
              <HighchartsReact
                highcharts={Highcharts}
                options={chartOptions}
                constructorType={"stockChart"}
                callback={this.afterChartCreated}
              />
            ) : (
              <Error fetchData={fetchReports} isLoading={isLoading} />
            )}
          </div>
        </div>
      </div>
    );
  }
}

export const ReportsChartConnected = connect(
  (state: AppState) => ({
    isLoading: state.reports.isLoading,
    hasError: state.reports.hasError,
    chartOptions: state.reports.chartOptions
  }),
  {
    fetchReports
  }
)(ReportsChart);

export default ReportsChart;
