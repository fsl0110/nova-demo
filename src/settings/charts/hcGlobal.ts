export const hcGlobal: Highcharts.Options = {
  chart: {
    height: 350,
    style: {
      fontFamily: "-apple-system, BlinkMacSystemFont, Segoe UI, Roboto, Oxygen"
    }
  },
  credits: {
    enabled: false
  },
  loading: {
    hideDuration: 500,
    showDuration: 300
  },
  title: {
    text: ""
  }
};
