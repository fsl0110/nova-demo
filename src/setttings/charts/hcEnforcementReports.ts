import moment from "moment";

export const hcEnforcementReports: Highcharts.Options = {
  yAxis: {
    title: {
      text: null
    }
  },
  xAxis: {
    type: "datetime",
    labels: {
      formatter: function() {
        return moment(this.value).format("MMM YYYY");
      }
    },
    minTickInterval: moment.duration(1, "month").asMilliseconds()
  },
  series: [{ type: "line", data: [null] }]
};
