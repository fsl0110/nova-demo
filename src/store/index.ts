import { createStore, applyMiddleware } from "redux";
import { composeWithDevTools } from "redux-devtools-extension";
import thunk from "redux-thunk";
import { reducer } from "./reducers";

export interface AppState {
  isLoading: {
    [key: string]: boolean;
  };
  hasError: {
    [key: string]: boolean;
  };
  data: {
    [key: string]: Array<(string | number)[]>;
  };
}

export type Path = string;

export const initialState: AppState = {
  isLoading: {
    REPORTS: false,
    INITIATORS: false
  },
  hasError: {
    REPORTS: false,
    INITIATORS: false
  },
  data: {
    REPORTS: [],
    INITIATORS: []
  }
};

export const middlewares = [thunk];

const store = createStore(
  reducer,
  initialState,
  composeWithDevTools(applyMiddleware(...middlewares))
);

export default store;
