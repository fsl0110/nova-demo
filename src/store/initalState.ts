export interface AppState {
  isLoading: {
    [key: string]: boolean;
  };
  hasError: {
    [key: string]: boolean;
  };
  data: {
    [key: string]: Array<(string | number)[]>;
  };
}

export const initialState: AppState = {
  isLoading: {
    REPORTS: false,
    INITIATORS: false
  },
  hasError: {
    REPORTS: false,
    INITIATORS: false
  },
  data: {
    REPORTS: [],
    INITIATORS: []
  }
};
