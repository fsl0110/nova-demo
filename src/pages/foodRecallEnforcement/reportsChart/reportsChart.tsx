import React, { PureComponent, HTMLAttributes } from "react";
import { connect } from "react-redux";
import produce from "immer";
import Highcharts from "highcharts/highstock";
import HighchartsReact from "highcharts-react-official";
import classNames from "classnames";
import { Search, Error } from "../../../components";
import { fetchReports } from "../../../store/actions";
import { AppState } from "../../../store/initalState";
import { hcGlobal } from "../../../utils/charts/hcGlobal";
import { hcEnforcementReports } from "../../../utils/charts/hcEnforcementReports";

require("highcharts-no-data-to-display")(Highcharts);

export interface StateProps {
  /** Is loading data */
  isLoading: boolean;
  /** Has an error */
  hasError: boolean;
  /** Fetched and formatted data**/
  data: Array<(string | number)[]>;
}

export interface DispatchProps {
  /** Action for fetching data. **/
  fetchReports: (term?: string) => void;
}

export interface OwnProps {}

export type Props = StateProps &
  DispatchProps &
  OwnProps &
  HTMLAttributes<HTMLDivElement>;

interface State {
  chartOptions: Highcharts.Options;
}

export class ReportsChart extends PureComponent<Props, State> {
  private chartInstance: any;

  constructor(props: Props) {
    super(props);

    this.afterChartCreated = this.afterChartCreated.bind(this);
    this.state = {
      chartOptions: { ...hcGlobal, ...hcEnforcementReports }
    };
  }

  afterChartCreated(chart: any) {
    this.chartInstance = chart;
  }

  static getDerivedStateFromProps(nextProps: Props, prevState: State) {
    if (prevState.chartOptions.series![0].data !== nextProps.data) {
      return produce(prevState, (draft: State) => {
        draft.chartOptions.series![0].data = nextProps.data;
      });
    }
    return null;
  }

  componentDidMount() {
    this.props.fetchReports("");
  }

  render() {
    const { isLoading, hasError, fetchReports, className } = this.props;

    if (!hasError && this.chartInstance && isLoading) {
      this.chartInstance.showLoading();
    }
    if (!hasError && this.chartInstance && !isLoading) {
      this.chartInstance.hideLoading();
    }

    return (
      <div className={classNames(className)}>
        <Search fetchData={fetchReports} isLoading={isLoading} />
        <div>
          <h3 className="horizontal-center">Enforcement reports over time</h3>
          <div className="chart-wrapper">
            {!hasError ? (
              <HighchartsReact
                highcharts={Highcharts}
                options={this.state.chartOptions}
                constructorType={"stockChart"}
                callback={this.afterChartCreated}
              />
            ) : (
              <Error fetchData={fetchReports} isLoading={isLoading} />
            )}
          </div>
        </div>
      </div>
    );
  }
}

export const ReportsChartConnected = connect(
  (state: AppState, ownProps: OwnProps): StateProps => ({
    isLoading: state.isLoading.REPORTS,
    hasError: state.hasError.REPORTS,
    data: state.data.REPORTS
  }),
  {
    fetchReports
  }
)(ReportsChart);

export default ReportsChart;
