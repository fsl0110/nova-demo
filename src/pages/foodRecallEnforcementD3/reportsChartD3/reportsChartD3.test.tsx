import React from "react";
import { shallow, render } from "enzyme";
import toJson from "enzyme-to-json";
import { ReportsChartConnected, ReportsChart, Props } from "./reportsChart";
import { findByTestAttr, testStore } from "../../../utils/tests";
import { initialState } from "../../../store/initalState";
import { Provider } from "react-redux";
import { hcGlobal } from "../../../utils/charts/hcGlobal";
import { hcEnforcementReports } from "../../../utils/charts/hcEnforcementReports";

describe("Reports snapshots", () => {
  const testCases: Array<[string, Props]> = [
    [
      "default",
      {
        fetchReports: jest.fn(),
        isLoading: false,
        hasError: false,
        data: [["123456", 100], ["123456", 50], ["123456", 25]]
      }
    ],
    [
      "is loading",
      {
        fetchReports: jest.fn(),
        isLoading: true,
        hasError: false,
        data: [["123456", 100], ["123456", 50], ["T123456", 25]]
      }
    ],
    [
      "has error",
      {
        fetchReports: jest.fn(),
        isLoading: false,
        hasError: true,
        data: [["123456", 100], ["123456", 50], ["T123456", 25]]
      }
    ]
  ];

  testCases.forEach(([item, options]) => {
    it(item, () => {
      expect(toJson(shallow(<ReportsChart {...options} />))).toMatchSnapshot();
    });
  });
});

describe("<ReportsConnected>", () => {
  let store: any;
  let wrapper: any;
  const testProps: any = {
    fetchData: jest.fn(),
    isLoading: false,
    hasError: false,
    chartOptions: { ...hcGlobal, ...hcEnforcementReports }
  };

  beforeEach(() => {
    store = testStore(initialState);
    wrapper = shallow(
      <Provider store={store}>
        <ReportsChartConnected {...testProps} />
      </Provider>
    ).dive();
  });

  it("Test connected props", () => {
    expect(wrapper.props().isLoading).toBe(false);
    expect(wrapper.props().hasError).toBe(false);
    expect(wrapper.props().chartOptions).toBe(testProps.chartOptions);
  });
});

describe("<Reports>", () => {
  let wrapper: any;
  const testProps: any = {
    fetchData: jest.fn(),
    isLoading: false,
    hasError: false,
    chartOptions: { ...hcGlobal, ...hcEnforcementReports }
  };

  beforeEach(() => {
    wrapper = render(<ReportsChart {...testProps} />);
  });

  /*   it("Render Search", () => {
    const search = findByTestAttr(wrapper, "search");
    expect(search.length).toBe(1);
  }); */

  /*   it("Render ChartHighstock", () => {
    const chart = findByTestAttr(wrapper, "chart");
    expect(chart.length).toBe(1);
  }); */
});
