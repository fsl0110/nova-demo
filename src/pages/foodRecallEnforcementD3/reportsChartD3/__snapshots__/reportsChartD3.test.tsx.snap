// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders toMatchSnapshot 1`] = `
<div
  className=""
>
  ReportsChartD3
</div>
`;
